include/sys/arc_impl.h

```diff
@@ -174,6 +178,7 @@ typedef struct l1arc_buf_hdr {
 	zfs_refcount_t		b_refcnt;
 
 	arc_callback_t		*b_acb;
+	arc_callback_t		**b_acbp;
```

Comment:
Why do we need two accesses to the callback?

Reply:
To efficiently implement FIFO to keep original order.  First pointers points to the first element, second -- to the acb_next of the last element.

Reply:
Is the ordering critical here? Just wondering about preserving the ordering vs the additional memory requirement for each l1hdr.

Reply:
I think it may be. In many benchmarks I saw that requests arriving in order from different threads and blocked on the same indirect block are reaching the data in wrong order, confusing speculative prefetcher.  Some time ago I made prefetcher to do speculations for read before reading indirect blocks, that significantly helped reads, but for writes it appeared difficult. I hope this change should if not fix the problem, but may be improve it a bit.

Reply:
I see. It is probably more efficient to make `arc_callback_t` have an extra pointer for the previous element so that we can process the callback list in reverse order in `arc_read_done`. This would mean that the memory penalty would only apply when we have a read in progress and not to every l1hdr and gives you the ordering you need. Would that suffice?

Reply:
It could be cheaper for the header, indeed.  I'll think about it closer tomorrow morning.  But if we care so much about this structure, we could probably save much more by hiding b_freeze_* under ZFS_DEBUG.

Reply:
Implemented.  First traversal left reversed to not traverse extra time.  Hope there will be no significant downsides.  The callbacks though should be called in right order, that should be more important.